Update region.py in skymatch

mcara opened the pull request:
Closes #5304
Closes #6439
Resolves [JP-1670](https://jira.stsci.edu/browse/JP-1670)

**Description**

Fixes issues in `skymatch` reported in #5304 and #6439 which are due to old `region.py` file used in the pipeline compared to the one in `stsci.skypac` which includes improvement and bug fixes that address the crashes reported in the above mentioned issues.

Bug fixes applied in `stsci.skypac` that are being ported to `jwst.skymatch.region` are:

- https://github.com/spacetelescope/stsci.skypac/pull/55
- https://github.com/spacetelescope/stsci.skypac/pull/52

Checklist
- [ ] Tests
- [ ] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

nden commented on jwst/skymatch/region.py:
```diff
@@ -1,8 +1,9 @@
-:Authors: Nadezhda Dencheva, Mihai Cara (contact: help@stsci.edu)
+:Authors: Nadezhda Dencheva, Mihai Cara
 
+:License: :doc:`LICENSE`
```
I would remove the authors and license to be consistent with the rest of the code in jwst.

nden commented on jwst/skymatch/region.py:
```diff
@@ -25,7 +26,6 @@
 
 __all__ = ['Region', 'Edge', 'Polygon']
 __taskname__ = 'region'
```
Is `__taskname__` left over from teal? If so remove it.

nden commented on jwst/skymatch/region.py:
```diff
@@ -36,18 +36,17 @@ def __str__(self):
 
 
-class Region():
+class Region(object):
```
```suggestion
class Region:
```

nden commented on jwst/skymatch/region.py:
```diff
@@ -36,18 +36,17 @@ def __str__(self):
     Parameters
     -------------
-    rid : int or string
+    rid: int or string
```
Not sure why this change but the entire docstring needs reformatting.
`---` should be the length of `Parameters` string
a space between `rid` and `:` is necessary (here and in all other docstrings)
the type should be `int or str`

nden commented on jwst/skymatch/region.py:
```diff
@@ -36,18 +36,17 @@ def __str__(self):
+    rid: int or string
         region ID
-    coordinate_system : astropy.wcs.CoordinateSystem instance or a string
+    coordinate_system: astropy.wcs.CoordinateSystem instance or a string
```
```suggestion
    coordinate_system : `~astropy.wcs.CoordinateSystem~ instance or str
```

nden commented on jwst/skymatch/region.py:
```diff
@@ -76,13 +75,14 @@ def scan(self, mask):
         Returns
         -------
-        mask : array where the value of the elements is the region ID or 0 (for
+        mask: array where the value of the elements is the region ID or 0 (for
```
```suggestion
        mask : ndarray 
```

nden replied:
And move the explanation on the next line

nden commented on jwst/skymatch/region.py:
```diff
@@ -93,18 +93,19 @@ class Polygon(Region):
     Parameters
     ----------
-    rid : string
+    rid: string
```
```suggestion
    rid : str
```

nden commented on jwst/skymatch/region.py:
```diff
@@ -223,26 +238,30 @@ def scan(self, data):
-            ysh = y + self._shifty
 
-            if ysh < 0 or ysh >= ny:
+            if y + self._shifty < 0 or y + self._shifty >= ny:
```
Computing `ysh` twice is better?

nden commented on jwst/skymatch/region.py:
```diff
@@ -275,21 +294,24 @@ def __contains__(self, px):
 
 
-class Edge():
+class Edge(object):
```
```suggestion
class Edge:
```

nden replied:
Use new style classes

nden commented:
I did a partial review and stopped. The changes are a mixture of style and code changes and it's hard to follow what and why is changed. I suggest splitting these into two PRs, one that fixes the problem and one for style improvements.

mcara commented:
@nden I have removed stylistic changes and left only bug-fix-related changes (+ one change with moving `int()` to a different line). I believe this should be much easier to review (see what the fix is).

I have also updated original/first comment to include links to the `stsci.skypac` PRs that introduced these fixes in `stsci.skypac`.

hbushouse reviewed (approved):
Looks good.

nden reviewed (approved):
Looks good. Thanks for splitting the bug fix into a separate PR.

nden commented:
Since the original code came from gwcs, do you think the fixes need to be applied there as well?

jemorrison commented:
Looks good to me

hbushouse commented:
Are we all OK with merging this now?

mcara commented:
> Since the original code came from gwcs, do you think the fixes need to be applied there as well?

Oh, I missed this message (due to my laptop troubles). I will investigate and open an issue if necessary.

@hbushouse It is OK to merge, IMO

mcara commented:
@nden The code from `gwcs`' `region.py` is way behind the one from `jwst`. It is probably possible to simply copy/paste the one from here into `gwcs`.
